File: frontend/features/battle/engine/battleLog.ts
import type {
  MainStatusCondition,
  StatName,
  VolatileStatusCondition,
} from "../types";

function getStatName(stat: StatName): string {
  switch (stat) {
    case "attack":
      return "こうげき";
    case "defense":
      return "ぼうぎょ";
    case "speed":
      return "すばやさ";
    case "accuracy":
      return "命中";
    case "evasion":
      return "回避";
  }
}

function getStatusName(
  condition: MainStatusCondition | VolatileStatusCondition
): string {
  switch (condition) {
    case "poison":
      return "どく";
    case "bad-poison":
      return "もうどく";
    case "burn":
      return "やけど";
    case "paralysis":
      return "まひ";
    case "sleep":
      return "ねむり";
    case "freeze":
      return "こおり";
    case "confusion":
      return "こんらん";
  }
}

export function createDamageLog(
  monsterName: string,
  damage: number
): string {
  return `${monsterName} に ${damage} ダメージ！`;
}

export function createCriticalLog(): string {
  return "急所に当たった！";
}

export function createGuardLog(monsterName: string): string {
  return `${monsterName} は身を守っている！`;
}

export function createGuardSuccessLog(monsterName: string): string {
  return `${monsterName} は攻撃を防いだ！`;
}

export function createStatChangeLog(
  monsterName: string,
  stat: StatName,
  stages: number
): string {
  const direction = stages > 0 ? "上がった" : "下がった";

  return `${monsterName} の${getStatName(stat)}が${direction}！`;
}

export function createAlreadyMainStatusLog(
  monsterName: string
): string {
  return `${monsterName} はすでに状態異常になっている！`;
}

export function createAlreadyVolatileStatusLog(
  monsterName: string
): string {
  return `${monsterName} はすでに状態変化している！`;
}

export function createMainStatusLog(
  monsterName: string,
  condition: MainStatusCondition
): string {
  return `${monsterName} は${getStatusName(condition)}状態になった！`;
}

export function createVolatileStatusLog(
  monsterName: string,
  condition: VolatileStatusCondition
): string {
  return `${monsterName} は${getStatusName(condition)}した！`;
}

File: backend/internal/battle/battlelog.go
package battle

import "strconv"

// getStatName はfeatures/battle/engine/battleLog.tsのgetStatNameの移植。
func getStatName(stat StatName) string {
	switch stat {
	case StatAttack:
		return "こうげき"
	case StatDefense:
		return "ぼうぎょ"
	case StatSpeed:
		return "すばやさ"
	case StatAccuracy:
		return "命中"
	case StatEvasion:
		return "回避"
	}
	return string(stat)
}

// getStatusName は本状態異常・状態変化どちらの文字列も受け付ける
// （TS側は2箇所に似た関数があるが、ここでは1つにまとめている）。
func getStatusName(condition string) string {
	switch MainStatusCondition(condition) {
	case StatusPoison:
		return "どく"
	case StatusBadPoison:
		return "もうどく"
	case StatusBurn:
		return "やけど"
	case StatusParalysis:
		return "まひ"
	case StatusSleep:
		return "ねむり"
	case StatusFreeze:
		return "こおり"
	}

	if VolatileStatusCondition(condition) == StatusConfusion {
		return "こんらん"
	}

	return condition
}

func createDamageLog(monsterName string, damage int) string {
	return monsterName + " に " + strconv.Itoa(damage) + " ダメージ！"
}

func createCriticalLog() string {
	return "急所に当たった！"
}

func createGuardLog(monsterName string) string {
	return monsterName + " は身を守っている！"
}

func createGuardSuccessLog(monsterName string) string {
	return monsterName + " は攻撃を防いだ！"
}

func createStatChangeLog(monsterName string, stat StatName, stages int) string {
	direction := "下がった"
	if stages > 0 {
		direction = "上がった"
	}

	return monsterName + " の" + getStatName(stat) + "が" + direction + "！"
}

func createAlreadyMainStatusLog(monsterName string) string {
	return monsterName + " はすでに状態異常になっている！"
}

func createAlreadyVolatileStatusLog(monsterName string) string {
	return monsterName + " はすでに状態変化している！"
}

func createMainStatusLog(monsterName string, condition MainStatusCondition) string {
	return monsterName + " は" + getStatusName(string(condition)) + "状態になった！"
}

func createVolatileStatusLog(monsterName string, condition VolatileStatusCondition) string {
	return monsterName + " は" + getStatusName(string(condition)) + "した！"
}
